packages/ui/src/components/Typography/Typography.tsx

```diff
@@ -0,0 +1,48 @@
+import React, { ReactNode, ReactText } from 'react';
+import { Box, forwardRef, HTMLChakraProps } from '@chakra-ui/react';
+
+type TypographyVariant = 'span' | 'div' | 'p' | 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6';
+
+type TypographyFontSize = 'xs' | 'sm' | 'md' | 'lg' | 'xl' | '2xl' | '3xl' | '4xl' | '5xl' | '6xl' | '7xl';
+
+type TypographyFontWeight = 'regular' | 'medium';
+
+type RemoveCommonValues<Type, TOmit> = {
+  [Property in keyof Type]: TOmit extends Record<Property, infer U> ? Exclude<Type[Property], U> : Type[Property];
+};
+
+type Id<Type> = Record<string, unknown> & { [P in keyof Type]: Type[P] };
+type ConditionalProps<Type, TKey extends keyof TCase, TCase extends Partial<Type>> =
+  | Id<Omit<Type, keyof TCase> & TCase>
+  | Id<RemoveCommonValues<Type, Pick<TCase, TKey>>>;
+
+export interface TypographyProps extends HTMLChakraProps<'div'> {
+  children: ReactText | ReactText[] | ReactNode | ReactNode[];
+  as?: TypographyVariant | 'a';
+  size?: TypographyFontSize;
+  weight?: TypographyFontWeight;
+}
+export interface TypographyPropsForAnchor extends TypographyProps {
+  href: string;
+  as: 'a';
+}
+
+export type IsAnchor = ConditionalProps<TypographyProps, 'as', TypographyPropsForAnchor>;
```

Comment:
I think you should save this type as a gist, as it might be useful for some components.
Saying so, there is no need to use it here as the previous solution worked really well. We don't want to require the `href` property as sometimes it would be provided by the parent component (e.g. [Next Link](https://nextjs.org/docs/api-reference/next/link)). I checked and even for the previous solution, I have IntelliSense for `href` and `target` props when using `as="a"` prop.

Reply:
Should I change the component type to the previous one here,   
  ```
children: ReactNode;
  as?: TypographyVariant | 'a';
  size?: TypographyFontSize;
  weight?: TypographyFontWeight;
``` and add optional href as string?

Reply:
Something like this
```
type TypographyVariant = 'span' | 'div' | 'a' | 'p' | 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6';

export interface TypographyProps extends HTMLChakraProps<'div'> {
  children: ReactNode;
  as?: TypographyVariant;
  size?: TypographyFontSize;
  weight?: TypographyFontWeight;
}
```

There is no need to specify the `href` property. As you extend `HTMLChakraProps` this component would have `href` in autosuggestions if someone provide `as="a"` in props.